Integrate npm package publish

jellespijker opened the pull request:
Publish a NPM package from a conan recipe.

This is based on `main` from the work done on https://github.com/Ultimaker/CuraEngine/pull/2183

Renamed the `conan-package.yml` workflow to `package.yml` since it is now also used for NPM packages

Added a `conf` to the recipe, which is used by the conan npm generator to generate a `package.json`. NPM packages are created on pushes to `main`, `NP-*` and `5.10` release branch and they will end up here: https://github.com/Ultimaker/CuraEngine/pkgs/npm/curaenginejs

_Yes, the hardcoded `5.10` hurts_. 
_Yes, the replacement of the `+` to `-` in the version of the `package.json` hurts_, but NPM kept "sanitizing" the build-meta data away and the version needs to be unique.

See also:
- https://github.com/Ultimaker/cura-workflows/pull/32 Which adds a npm-package workflow, and allows for conan config test branches
- https://github.com/Ultimaker/conan-config/pull/20 Which adds a npm generator
- https://github.com/Ultimaker/conan-ultimaker-index/pull/18 which adds a python requires that handles package_json in the conf of a recipe

# Checklist:

- [X] My code follows the style guidelines of this project as described in [UltiMaker Meta](https://github.com/Ultimaker/Meta)
- [X] I have read the [Contribution guide](https://github.com/Ultimaker/Cura/blob/main/contributing.md)
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have uploaded any files required to test this change

jellespijker committed:
Integrate npm package publish

Use the new workflows for Conan v2 to build and publish
and npm package

Contribute to NP-637

rburema reviewed (approved):
lgtm, but I do have a few suggestions

jellespijker commented on conanfile.py:
```diff
@@ -202,3 +204,22 @@ def package_info(self):
         self.conf_info.define_path("user.curaengine:curaengine",
                                    os.path.join(self.package_folder, "bin", f"CuraEngine{ext}"))
+        if self.settings.os == "Emscripten":
+            package_json = {
+                "name": f"@ultimaker/{self.name.lower()}js",
+                "version": f"{str(self.version).replace('+', '-')}",  # npm will otherwise 'sanitize' the version number
+                "description": f"JavaScript / TypeScript bindings for {self.name}, a {self.description}",
+                "main": "bin/CuraEngine.js",
+                "repository": {
+                    "type": "git",
+                    "url": self.url
+                },
+                "author": self.author,
+                "license": self.license,
+                "keywords": self.topics,
+                "files": [
+                    "bin",
+                    "package.json"
+                ]
+            }
```
replaced it with a python requires: https://github.com/Ultimaker/conan-ultimaker-index/pull/18

jellespijker committed:
Add npmpackage dependency and delegate package.json setup

The npmpackage Python requirement was introduced to handle package.json configuration for Emscripten builds. This simplifies the code by delegating the package.json setup to the npmpackage module, reducing redundancy and maintaining cleaner logic.

Contribute to NP-637

jellespijker committed:
Enforce minimum compiler versions and C++20 support

Added checks to validate minimum compiler versions and enforce C++20 compatibility for the package. These include GCC 12, Clang 14, Apple Clang 13, MSVC 191, and Visual Studio 17. This ensures compatibility with the required C++ features, preventing builds with unsupported compilers.

Contribute to NP-637

jellespijker committed:
bogus commit

Contribute to NP-637

jellespijker committed:
implement review comments

Contribute to NP-637

wawanbreton reviewed (approved):
One remark, looks good otherwise
